lib/WasmReader/WasmByteCodeGenerator.cpp

```diff
@@ -1182,62 +1123,59 @@ WasmBytecodeGenerator::EmitReturnExpr(EmitInfo* explicitRetInfo)
+        m_writer->AsmReg1(Js::OpCodeAsmJs::LdUndef, 0);
     }
     else
     {
         EmitInfo retExprInfo = explicitRetInfo ? *explicitRetInfo : PopEvalStack();
-
-        if (m_funcInfo->GetResultType() != retExprInfo.type)
+        if (retExprInfo.type != WasmTypes::Any)
         {
-            throw WasmCompilationException(_u("Result type must match return type"));
-        }
+            if (m_funcInfo->GetResultType() != retExprInfo.type)
+            {
+                throw WasmCompilationException(_u("Result type must match return type"));
+            }
 
-        Js::OpCodeAsmJs retOp = GetReturnOp(retExprInfo.type);
-        m_writer.Conv(retOp, 0, retExprInfo.location);
-        ReleaseLocation(&retExprInfo);
+            Js::OpCodeAsmJs retOp = GetReturnOp(retExprInfo.type);
+            m_writer->Conv(retOp, 0, retExprInfo.location);
+            ReleaseLocation(&retExprInfo);
+        }
     }
-    m_writer.AsmBr(m_funcInfo->GetExitLabel());
+    m_writer->AsmBr(m_funcInfo->GetExitLabel());
 
     SetUnreachableState(true);
 }
 
 EmitInfo
 WasmBytecodeGenerator::EmitSelect()
 {
-    EmitInfo conditionInfo = PopEvalStack();
-    if (conditionInfo.type != WasmTypes::I32)
-    {
-        throw WasmCompilationException(_u("select condition must have I32 type"));
-    }
+    EmitInfo conditionInfo = PopEvalStack(WasmTypes::I32, _u("select condition must have i32 type"));
 
-    Js::ByteCodeLabel falseLabel = m_writer.DefineLabel();
-    Js::ByteCodeLabel doneLabel = m_writer.DefineLabel();
+    Js::ByteCodeLabel falseLabel = m_writer->DefineLabel();
+    Js::ByteCodeLabel doneLabel = m_writer->DefineLabel();
 
-    m_writer.AsmBrReg1(Js::OpCodeAsmJs::BrFalse_Int, falseLabel, conditionInfo.location);
+    m_writer->AsmBrReg1(Js::OpCodeAsmJs::BrFalse_Int, falseLabel, conditionInfo.location);
     ReleaseLocation(&conditionInfo);
 
     EmitInfo falseInfo = PopEvalStack();
-    EmitInfo trueInfo = PopEvalStack();
+    EmitInfo trueInfo = PopEvalStack(falseInfo.type, _u("select operands must both have same type"));
 
     // Refresh the lifetime of the true location
-    m_writer.AsmReg2(GetLoadOp(trueInfo.type), trueInfo.location, trueInfo.location);
+    m_writer->AsmReg2(GetLoadOp(trueInfo.type), trueInfo.location, trueInfo.location);
 
-    m_writer.AsmBr(doneLabel);
-    m_writer.MarkAsmJsLabel(falseLabel);
-
-    if (trueInfo.type != falseInfo.type)
-    {
-        throw WasmCompilationException(_u("select operands must both have same type"));
-    }
+    m_writer->AsmBr(doneLabel);
+    m_writer->MarkAsmJsLabel(falseLabel);
 
     Js::OpCodeAsmJs op = GetLoadOp(trueInfo.type);
 
-    m_writer.AsmReg2(op, trueInfo.location, falseInfo.location);
+    m_writer->AsmReg2(op, trueInfo.location, falseInfo.location);
     ReleaseLocation(&falseInfo);
 
-    m_writer.MarkAsmJsLabel(doneLabel);
-
+    m_writer->MarkAsmJsLabel(doneLabel);
+    
+    if (trueInfo.type == WasmTypes::Any && falseInfo.type != WasmTypes::Any)
```

Comment:
Actually, in a select we have to evaluate both the true and the false result since it's a stack machine.
By the time we reach the select op we've already became unreachable therefore, the select is not executed.
The code really is
```js
var true = return;
var false = 5;
var selected = cond ? true : false;
```

This is remnant of when I was trying to emit valid bytecode in unreachable state, but with EmptyWasmBYteCodeWriter this became useless.

It make me think however that we probably should convert the select op to
```js
var selected;
if (cond) {
  selected = "true expr"; 
} else {
  selected = "false expr";
}
```
if neither expr has side effects